contracts/src/bridge/SequencerInbox.sol

```diff
@@ -146,20 +144,19 @@ contract SequencerInbox is DelegateCallAware, GasRefundEnabled, ISequencerInbox
         // solhint-disable-next-line avoid-tx-origin
         if (msg.sender != tx.origin) revert NotOrigin();
         if (!isBatchPoster[msg.sender]) revert NotBatchPoster();
-        if (inboxAccs.length != sequenceNumber) revert BadSequencerNumber();
         (bytes32 dataHash, TimeBounds memory timeBounds) = formDataHash(
             data,
             afterDelayedMessagesRead
         );
-        (bytes32 beforeAcc, bytes32 delayedAcc, bytes32 afterAcc) = addSequencerL2BatchImpl(
-            dataHash,
-            afterDelayedMessagesRead
-        );
-        if (data.length > 0) {
-            _reportBatchSpending(dataHash, sequenceNumber);
-        }
+        (
+            uint256 seqMessageCount,
+            bytes32 beforeAcc,
+            bytes32 delayedAcc,
+            bytes32 afterAcc
+        ) = addSequencerL2BatchImpl(dataHash, afterDelayedMessagesRead, data.length);
+        if (seqMessageCount != sequenceNumber) revert BadSequencerNumber();
```

Comment:
It is a bit unfortunate that this would revert after a lot of gas has been spent, but I think it's an acceptable tradeoff.

Reply:
We could query the sequencer acc's length at the start of the function, should be little gas overhead.
But I prefer this approach since we optimise the happy path cost